File: plugins/org.eclipse.tm.tcf.core/src/org/eclipse/tm/internal/tcf/services/remote/DiagnosticsProxy.java
/*******************************************************************************
 * Copyright (c) 2007, 2008 Wind River Systems, Inc. and others.
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html 
 *  
 * Contributors:
 *     Wind River Systems - initial API and implementation
 *******************************************************************************/
package org.eclipse.tm.internal.tcf.services.remote;

import java.math.BigDecimal;
import java.util.Collection;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.services.IDiagnostics;


public class DiagnosticsProxy implements IDiagnostics {
    
    private final IChannel channel;
    
    private static class Symbol implements ISymbol {
        
        private final Map<String,Object> props;
        
        Symbol(Map<String,Object> props) {
            this.props = props;
        }

        public String getSectionName() {
            return (String)props.get("Section");
        }

        public Number getValue() {
            return (Number)props.get("Value");
        }

        public boolean isAbs() {
            Boolean b = (Boolean)props.get("Abs");
            return b != null && b.booleanValue();
        }

        public boolean isCommon() {
            String s = (String)props.get("Storage");
            return s != null && s.equals("COMMON");
        }

        public boolean isGlobal() {
            String s = (String)props.get("Storage");
            return s != null && s.equals("GLOBAL");
        }

        public boolean isLocal() {
            String s = (String)props.get("Storage");
            return s != null && s.equals("LOCAL");
        }

        public boolean isUndef() {
            String s = (String)props.get("Storage");
            return s != null && s.equals("UNDEF");
        }
    }
    
    public DiagnosticsProxy(IChannel channel) {
        this.channel = channel;
    }

    public String getName() {
        return NAME;
    }

    public IToken echo(String s, final DoneEcho done) {
        return new Command(channel, this, "echo", new Object[]{ s }) {
            @Override
            public void done(Exception error, Object[] args) {
                String str = null;
                if (error == null) {
                    assert args.length == 1;
                    str = (String)args[0];
                }
                done.doneEcho(token, error, str);
            }
        }.token;
    }

    public IToken echoFP(BigDecimal n, final DoneEchoFP done) {
        return new Command(channel, this, "echoFP", new Object[]{ n }) {
            @Override
            public void done(Exception error, Object[] args) {
                BigDecimal n = null;
                if (error == null) {
                    assert args.length == 1;
                    Number x = (Number)args[0];
                    if (x instanceof BigDecimal) {
                        n = (BigDecimal)x;
                    }
                    else {
                        n = new BigDecimal(x.toString());
                    }
                }
                done.doneEchoFP(token, error, n);
            }
        }.token;
    }

    public IToken getTestList(final DoneGetTestList done) {
        return new Command(channel, this, "getTestList", null) {
            @Override
            public void done(Exception error, Object[] args) {
                String[] arr = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    arr = toStringArray(args[1]);
                }
                done.doneGetTestList(token, error, arr);
            }
        }.token;
    }

    public IToken runTest(String s, final DoneRunTest done) {
        return new Command(channel, this, "runTest", new Object[]{ s }) {
            @Override
            public void done(Exception error, Object[] args) {
                String str = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    str = (String)args[1];
                }
                done.doneRunTest(token, error, str);
            }
        }.token;
    }

    public IToken cancelTest(String s, final DoneCancelTest done) {
        return new Command(channel, this, "cancelTest", new Object[]{ s }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCancelTest(token, error);
            }
        }.token;
    }

    public IToken getSymbol(String context_id, String symbol_name, final DoneGetSymbol done) {
        return new Command(channel, this, "getSymbol", new Object[]{ context_id, symbol_name }) {
            @Override
            public void done(Exception error, Object[] args) {
                ISymbol sym = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    sym = toSymbol(args[1]);
                }
                done.doneGetSymbol(token, error, sym);
            }
        }.token;
    }

    public IToken createTestStreams(int inp_buf_size, int out_buf_size, final DoneCreateTestStreams done) {
        return new Command(channel, this, "createTestStreams", new Object[]{ inp_buf_size, out_buf_size }) {
            @Override
            public void done(Exception error, Object[] args) {
                String inp_id = null;
                String out_id = null;
                if (error == null) {
                    assert args.length == 3;
                    error = toError(args[0]);
                    inp_id = (String)args[1];
                    out_id = (String)args[2];
                }
                done.doneCreateTestStreams(token, error, inp_id, out_id);
            }
        }.token;
    }

    public IToken disposeTestStream(String id, final DoneDisposeTestStream done) {
        return new Command(channel, this, "disposeTestStream", new Object[]{ id }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneDisposeTestStream(token, error);
            }
        }.token;
    }

    public IToken not_implemented_command(final DoneNotImplementedCommand done) {
        return new Command(channel, this, "not implemented command", null) {
            @Override
            public void done(Exception error, Object[] args) {
                done.doneNotImplementedCommand(token, error);
            }
        }.token;
    }

    @SuppressWarnings("unchecked")
    private String[] toStringArray(Object o) {
        if (o == null) return null;
        Collection<String> c = (Collection<String>)o;
        return (String[])c.toArray(new String[c.size()]);
    }
    
    @SuppressWarnings("unchecked")
    private ISymbol toSymbol(Object o) {
        if (o == null) return null;
        return new Symbol((Map<String,Object>)o);
    }
}

File: org.eclipse.tcf/python/src/tcf/services/remote/DiagnosticsProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

import time
from tcf import errors
from tcf.services import diagnostics
from tcf.channel.Command import Command

class DiagnosticsProxy(diagnostics.DiagnosticsService):
    def __init__(self, channel):
        self.channel = channel

    def echo(self, s, done):
        done = self._makeCallback(done)
        service = self
        class EchoCommand(Command):
            def __init__(self):
                super(EchoCommand, self).__init__(service.channel, service, "echo", (s,))
            def done(self, error, args):
                str = None
                if not error:
                    assert len(args) == 1
                    str = args[0]
                done.doneEcho(self.token, error, str)
        return EchoCommand().token

    def echoFP(self, n, done):
        done = self._makeCallback(done)
        service = self
        class EchoFPCommand(Command):
            def __init__(self):
                super(EchoFPCommand, self).__init__(service.channel, service, "echoFP", (n,))
            def done(self, error, args):
                n = None
                if not error:
                    assert len(args) == 1
                    n = args[0]
                done.doneEchoFP(self.token, error, n)
        return EchoFPCommand().token

    def echoERR(self, err, done):
        map = None
        if isinstance(err, errors.ErrorReport):
            map = err.getAttributes()
        else:
            map = {
                errors.ERROR_TIME : int(time.time() * 1000),
                errors.ERROR_CODE : errors.TCF_ERROR_OTHER,
                errors.ERROR_FORMAT : err.message
            }
        done = self._makeCallback(done)
        service = self
        class EchoERRCommand(Command):
            def __init__(self):
                super(EchoERRCommand, self).__init__(service.channel, service, "echoERR", (map,))
            def done(self, error, args):
                err = None
                str = None
                if not error:
                    assert len(args) == 2
                    err = self.toError(args[0])
                    str = args[1]
                done.doneEchoERR(self.token, error, err, str)
        return EchoERRCommand().token

    def getTestList(self, done):
        done = self._makeCallback(done)
        service = self
        class GetTestListCommand(Command):
            def __init__(self):
                super(GetTestListCommand, self).__init__(service.channel, service, "getTestList", None)
            def done(self, error, args):
                arr = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    arr = args[1]
                done.doneGetTestList(self.token, error, arr)
        return GetTestListCommand().token

    def runTest(self, s, done):
        done = self._makeCallback(done)
        service = self
        class RunTestCommand(Command):
            def __init__(self):
                super(RunTestCommand, self).__init__(service.channel, service, "runTest", (s,))
            def done(self, error, args):
                str = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    str = args[1]
                done.doneRunTest(self.token, error, str)
        return RunTestCommand().token

    def cancelTest(self, s, done):
        done = self._makeCallback(done)
        service = self
        class CancelTestCommand(Command):
            def __init__(self):
                super(CancelTestCommand, self).__init__(service.channel, service, "cancelTest", (s,))
            def done(self, error, args):
                if not error:
                    assert len(args) == 1
                    error = self.toError(args[0])
                done.doneCancelTest(self.token, error)
        return CancelTestCommand().token

    def getSymbol(self, context_id, symbol_name, done):
        done = self._makeCallback(done)
        service = self
        class GetSymbolCommand(Command):
            def __init__(self):
                super(GetSymbolCommand, self).__init__(service.channel, service, "getSymbol", (context_id, symbol_name))
            def done(self, error, args):
                sym = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    sym = _toSymbol(args[1])
                done.doneGetSymbol(self.token, error, sym)
        return GetSymbolCommand().token

    def createTestStreams(self, inp_buf_size, out_buf_size, done):
        done = self._makeCallback(done)
        service = self
        class CreateTestStreamsCommand(Command):
            def __init__(self):
                super(CreateTestStreamsCommand, self).__init__(service.channel, service, "createTestStreams", (inp_buf_size, out_buf_size))
            def done(self, error, args):
                inp_id = None
                out_id = None
                if not error:
                    assert len(args) == 3
                    error = self.toError(args[0])
                    inp_id = args[1]
                    out_id = args[2]
                done.doneCreateTestStreams(self.token, error, inp_id, out_id)
        return CreateTestStreamsCommand().token

    def disposeTestStream(self, id, done):
        done = self._makeCallback(done)
        service = self
        class DisposeTestStreamCommand(Command):
            def __init__(self):
                super(DisposeTestStreamCommand, self).__init__(service.channel, service, "disposeTestStream", (id,))
            def done(self, error, args):
                if not error:
                    assert len(args) == 1
                    error = self.toError(args[0])
                done.doneDisposeTestStream(self.token, error)
        return DisposeTestStreamCommand().token

    def not_implemented_command(self, done):
        done = self._makeCallback(done)
        service = self
        class NotImplementedCommand(Command):
            def __init__(self):
                super(NotImplementedCommand, self).__init__(service.channel, service, "not implemented command", None)
            def done(self, error, args):
                done.doneNotImplementedCommand(self.token, error)
        return NotImplementedCommand().token

def _toSymbol(o):
    if o is None: return None
    return diagnostics.Symbol(o)
